lib/rucio/core/transfer.py

```diff
@@ -972,68 +972,68 @@ def for_request(
         if source.rse.attributes.get(RseAttr.RESTRICTED_READ) and ctx.rws.account not in self.admin_accounts:
             return SKIP_SOURCE
 
 
 class SkipBlocklistedRSEs(SourceFilterStrategy):
 
-    def __init__(self, topology: "Topology"):
+    def __init__(self, topology: "Topology") -> None:
         super().__init__()
         self.topology = topology
 
-    def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[int | _SkipSource]":
+    def apply(self, ctx: RequestRankingContext, source: RequestSource) -> Optional[Union[int, _SkipSource]]:
         # Ignore blocklisted RSEs
         if not source.rse.columns['availability_read'] and not self.topology.ignore_availability:
             return SKIP_SOURCE
 
 
 class EnforceStagingBuffer(SourceFilterStrategy):
-    def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[int | _SkipSource]":
+    def apply(self, ctx: RequestRankingContext, source: RequestSource) -> Optional[Union[int, _SkipSource]]:
         # For staging requests, the staging_buffer attribute must be correctly set
         if ctx.rws.request_type == RequestType.STAGEIN and source.rse.attributes.get(RseAttr.STAGING_BUFFER) != ctx.rws.dest_rse.name:
             return SKIP_SOURCE
 
 
 class RestrictTapeSources(SourceFilterStrategy):
-    def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[int | _SkipSource]":
+    def apply(self, ctx: RequestRankingContext, source: RequestSource) -> Optional[Union[int, _SkipSource]]:
         # Ignore tape sources if they are not desired
         if source.rse.is_tape_or_staging_required() and not ctx.rws.attributes.get("allow_tape_source", True):
             return SKIP_SOURCE
 
 
 class HighestAdjustedRankingFirst(SourceRankingStrategy):
-    def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[int | _SkipSource]":
+    def apply(self, ctx: RequestRankingContext, source: RequestSource) -> Optional[Union[int, _SkipSource]]:
         source_ranking_penalty = 1 if source.rse.is_tape_or_staging_required() else 0
         return - source.ranking + source_ranking_penalty
 
 
 class PreferDiskOverTape(SourceRankingStrategy):
-    def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[int | _SkipSource]":
+    def apply(self, ctx: RequestRankingContext, source: RequestSource) -> Optional[Union[int, _SkipSource]]:
         return int(source.rse.is_tape_or_staging_required())  # rely on the fact that False < True
 
 
 class PathDistance(SourceRankingStrategy):
 
     class _RankingContext(RequestRankingContext):
-        def __init__(self, strategy: "SourceRankingStrategy", rws: "RequestWithSources", paths_for_rws: "Mapping[RseData, Sequence[DirectTransfer]]"):
+        def __init__(self, strategy: SourceRankingStrategy, rws: RequestWithSources, paths_for_rws: dict[RseData, list[DirectTransfer]]) -> None:
```

Comment:
I think this should be reverted, unless it was causing type errors before.

I think you made this change because you changed `build_or_return_cached` to always return `dict` instead of `Mapping` (which is correct, because that is the type returned and we should be as specific as possible with return types of private functions).

In the context of `_RankingContext`:  if you were to pass a `Mapping` for `paths_for_rws`, instead of specifically a `dict`, all usages of `_RankingContext` would still work, since we're only using the `get` method on `paths_for_rws`, and `get` exists on `Mapping`. So `_RankingContext` doesn't necessarily need a `dict` here, it can accept any `Mapping`.